docs/proposals/block-storage-time-series-deletion.md

```diff
@@ -0,0 +1,272 @@
+
+The tombstone will be stored in a single file per request: 
+
+- `/<tenantId>/series_deletion/requests/<request_id>.json`
+
+
+The schema of the JSON file is:
+
+
+```
+{
+  "requestId": <string>,
+  "startTime": <int>,
+  "endTime": <int>,
+  "creationTime": <int>,
+  "matchers": [
+    "<string matcher 1>",
+    .., 
+    "<string matcher n>"
+    ]
+  },
+  "userID": <string>,
+  "deleteRequestStatus": <received | buildingPlan | deleting | processed> 
+}
+```
+
+
+Pros:
+
+- Design is similar to the existing chunk storage deletion 
+    - Lots of code can be reused inside the purger component. 
+- Allows deletion and un-delete to be done in a single operation. 
+
+Cons:
+- Negative impact on query performance when there are active tombstones. As in the chunk storage implementation, all the queries made will have to be compared to the matchers contained in the active tombstone files. The impact on performance should be the same as the deletion would have with chunk storage.
+
+
+#### Invalidating cache
+
+Using block store, the different caches available are:
+- Index cache
+- Metadata cache
+- Chunks cache (stores the potentially to be deleted chunks of data) 
+- Query results cache (stores the potentially to be deleted data) 
+
+The filtering using tombstones is only requested by the querier. By using the purger, the querier filters out the data received from the ingestors and store-gateway. The cache not being processed through the querier needs to be invalidated to prevent deleted data from coming up in queries. There are two potential caches that could contain deleted data, the chunks cache, and the query results cache. 
+
+
+Firstly, the query results cache needs to be invalidated for each new delete request. This can be done using the same mechanism currently used for chunk storage by utilizing the cache generation numbers. For each tenant, their cache is prefixed with a cache generation number. This is already implemented into the middleware and would be easy to use for invalidating the cache. When the cache needs to be invalidated due to a delete or cancel delete request, the cache generation numbers would be increased (to the current timestamp), which would invalidate all the cache entries for a given tenant. The cache generation numbers are currently being stored in an Index table  (e.g. DynamoDB or Bigtable). One option for block store is to store a per tenant JSON file with the corresponding cache generation numbers. 
+
+Furthermore, since the chunks cache  is retrieved from the store gateway and passed to the querier, it will be filtered out like the rest of the time series data in the querier using the tombstones, with the mechanism described in the previous section. However, some issues may arise if the tombstone is deleted but the data to-be-deleted still exists in the chunks cache. This is why it is also best to use cache generation numbers for this cache in order to invalidate it the same way. 
```

Comment:
> This is why it is also best to use cache generation numbers for this cache in order to invalidate it the same way. 

I think it's undesiderable flushing the whole chunks cache each time a deletion request is made. An option may be keep filtering out tombstones at query time for a period of time after which the tombstone has been deleted by compactor, to give store-gateways enough time to pick up new blocks and drop outdated ones. I think this delay would be the `-compactor.deletion-delay` + `-blocks-storage.bucket-store.sync-interval` + some delta.

Reply:
That makes sense. To allow the filtering to keep taking place after the deletion has already taken place, I've added a new state in the deletion process called `syncing`. This state will last a fixed amount of time until the delay you mentioned has been completed.